slatedb/src/db/builder.rs

```diff
@@ -212,35 +226,47 @@ impl<P: Into<Path>> DbBuilder<P> {
         };
 
         // Setup object store with optional caching
-        let maybe_cached_object_store = match &self.settings.object_store_cache_options.root_folder
-        {
-            None => self.object_store.clone(),
-            Some(cache_root_folder) => {
-                let stats = Arc::new(CachedObjectStoreStats::new(stat_registry.as_ref()));
-                let cache_storage = Arc::new(FsCacheStorage::new(
-                    cache_root_folder.clone(),
-                    self.settings
-                        .object_store_cache_options
-                        .max_cache_size_bytes,
-                    self.settings.object_store_cache_options.scan_interval,
-                    stats.clone(),
-                ));
-
-                let cached_object_store = CachedObjectStore::new(
-                    self.object_store.clone(),
-                    cache_storage,
-                    self.settings.object_store_cache_options.part_size_bytes,
-                    stats.clone(),
-                )?;
-                cached_object_store.start_evictor().await;
-                cached_object_store
-            }
-        };
+        let maybe_cached_main_object_store =
+            match &self.settings.object_store_cache_options.root_folder {
+                None => self.main_object_store.clone(),
+                Some(cache_root_folder) => {
+                    let stats = Arc::new(CachedObjectStoreStats::new(stat_registry.as_ref()));
+                    let cache_storage = Arc::new(FsCacheStorage::new(
+                        cache_root_folder.clone(),
+                        self.settings
+                            .object_store_cache_options
+                            .max_cache_size_bytes,
+                        self.settings.object_store_cache_options.scan_interval,
+                        stats.clone(),
+                    ));
+
+                    let cached_main_object_store = CachedObjectStore::new(
+                        self.main_object_store.clone(),
+                        cache_storage,
+                        self.settings.object_store_cache_options.part_size_bytes,
+                        stats.clone(),
+                    )?;
+                    cached_main_object_store.start_evictor().await;
+                    cached_main_object_store
+                }
+            };
 
         // Setup the manifest store and load latest manifest
-        let manifest_store = Arc::new(ManifestStore::new(&path, maybe_cached_object_store.clone()));
+        let manifest_store = Arc::new(ManifestStore::new(
+            &path,
+            maybe_cached_main_object_store.clone(),
+        ));
         let latest_manifest = StoredManifest::try_load(manifest_store.clone()).await?;
 
+        // Validate WAL object store configuration
+        if let Some(latest_manifest) = &latest_manifest {
+            if latest_manifest.db_state().wal_object_store_uri != wal_object_store_uri {
```

Comment:
when we start to store the actual uri, is the plan to check for the empty string and allow it to be overwritten?

Reply:
my idea was just to fail, we can try to relax it a bit and allow overwriting an empty URI, but this will require updating all (?) the manifests, not sure how it will play with `DbReader` for instance

Reply:
... probably it's better to declare it as a transitional/unstable feature and just fail

Reply:
> my idea was just to fail, we can try to relax it a bit and allow overwriting an empty URI, but this will require updating all (?) the manifests, not sure how it will play with DbReader for instance

This should be fine. We'd just need to update the "current" manifest by writing a new version and then future clients would just work.

> ... probably it's better to declare it as a transitional/unstable feature and just fail

This seems fine just to keep the implementation simple